### ISSUE TITLE
[Light client] Slow down of header sync rate after several seconds

### ISSUE THREAD

**Issue by Tbaut** [paritytech/parity-ethereum#9880]
_Before filing a new issue, please **provide the following information**._

- **Parity Ethereum version**: v2.1.3-beta-
- **Operating system**: Windows and Linux
- **Installation**: binary
- **Fully synchronized**: no
- **Network**: kovan
- **Restarted**: yes

When syncing from the latest hardcoded header (`--light db kill` before syncing), the light client starts to download headers at a high rate (x000 hdr/s), and then it slows down quite substantially (x00 hdr/s). 

```bash
2018-11-07 17:07:14  Starting Parity-Ethereum/v2.1.3-beta-18ddd7c-20181015/x86_64-linux-gnu/rustc1.29.0
2018-11-07 17:07:14  Keys path /home/thib/.local/share/io.parity.ethereum/keys/kovan
2018-11-07 17:07:14  DB path /home/thib/.local/share/io.parity.ethereum/chains_light/kovan/db/9bf388941c25ea98
2018-11-07 17:07:14  Running in experimental Light Client mode.
2018-11-07 17:07:14  Debug API is not available in light client mode.
2018-11-07 17:07:14  Debug API is not available in light client mode.
2018-11-07 17:07:16  Imported #9094915 0x21ff…aaef (0.35 Mgas)
2018-11-07 17:07:17  Imported #9099267 0xded5…215f (0.36 Mgas) + another 255 header(s)
2018-11-07 17:07:19  Syncing #9106658 0xce5a…81f9  2553.6 hdr/s      0+16127 Qed  13/50 peers   10 MiB cache 12 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:07:19  Public node URL: enode://8c820b97320e978b841fe62fed9a6ac2baca0ff59abe40460d2fba7e062ebe3ec4ad48c815a9656d9b26ba3f593640bb4ac4347fcbf0362637770d724b96d7c2@192.168.1.159:30303
2018-11-07 17:07:20  Imported #9109507 0x577b…8203 (0.76 Mgas) + another 255 header(s)
2018-11-07 17:07:24  Syncing #9128846 0x0f91…00e3  4437.6 hdr/s      0+15615 Qed  14/50 peers   10 MiB cache 12 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:07:24  Imported #9128963 0x2414…23ee (0.28 Mgas) + another 255 header(s)
2018-11-07 17:07:29  Syncing #9145090 0x6e1b…4667  3248.8 hdr/s      0+    0 Qed  13/50 peers   10 MiB cache 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:07:30  Imported #9145091 0x9e44…c04d (0.23 Mgas)
2018-11-07 17:07:34  Syncing #9145858 0x67e2…0949   153.6 hdr/s      0+    0 Qed  13/50 peers   10 MiB cache 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:07:34  Imported #9145859 0x85fe…b4c8 (1.52 Mgas)
2018-11-07 17:07:39  Syncing #9146882 0x3200…d05b   204.8 hdr/s      0+    0 Qed  14/50 peers   10 MiB cache 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:07:40  Imported #9146883 0x81d9…c030 (0.62 Mgas)
2018-11-07 17:07:44  Syncing #9147394 0xa81b…208c   102.4 hdr/s      0+    0 Qed  16/50 peers   10 MiB cache 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
```
if I `ctrl-C` and relauch, the rate is quite high again for a couple seconds, and then goes back down.
```bash
2018-11-07 17:23:56  Starting Parity-Ethereum/v2.1.3-beta-18ddd7c-20181015/x86_64-linux-gnu/rustc1.29.0
2018-11-07 17:23:56  Keys path /home/thib/.local/share/io.parity.ethereum/keys/kovan
2018-11-07 17:23:56  DB path /home/thib/.local/share/io.parity.ethereum/chains_light/kovan/db/9bf388941c25ea98
2018-11-07 17:23:56  Running in experimental Light Client mode.
2018-11-07 17:23:57  Debug API is not available in light client mode.
2018-11-07 17:23:57  Debug API is not available in light client mode.
2018-11-07 17:23:59  Imported #9160196 0x24d1…7434 (1.59 Mgas)
2018-11-07 17:24:00  Imported #9161524 0x3d7b…d4ad (0.31 Mgas) + another 255 header(s)
2018-11-07 17:24:02  Syncing #9165110 0xe396…6442   983.0 hdr/s      0+29775 Qed  10/50 peers   5 MiB cache 23 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:24:02  Public node URL: enode://8c820b97320e978b841fe62fed9a6ac2baca0ff59abe40460d2fba7e062ebe3ec4ad48c815a9656d9b26ba3f593640bb4ac4347fcbf0362637770d724b96d7c2@192.168.1.159:30303
2018-11-07 17:24:07  Syncing #9181222 0xcf8c…ecef  3222.4 hdr/s      0+29132 Qed   9/50 peers   10 MiB cache 22 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:24:12  Syncing #9198132 0x62b4…eb49  3382.0 hdr/s      0+12239 Qed  10/50 peers   10 MiB cache 9 MiB queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:24:17  Syncing #9210627 0x96c5…eecb  2499.0 hdr/s      0+    0 Qed   9/50 peers   10 MiB cache 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs
2018-11-07 17:24:17  Imported #9210628 0x7c99…41d2 (0.02 Mgas)
2018-11-07 17:24:22  Syncing #9210883 0x4786…45c3    51.2 hdr/s      0+    0 Qed   9/50 peers   10 MiB cache 0 bytes queue  RPC:  0 conn,    0 req/s,    0 µs

```
Note that the light client alternates between `Imported` and `Syncing`, while the top of the chain is still very far (9.3M as of writing, and the client is at 9.1M)

We originally thought that it was due to RPC sent from Fether, but this also happens without fether attached.

**Comment by tomusdrw** [comment]
It's the light client built-in rate limiting. You run out of credits with the peers you are connected to, since it's not persistent after restart you start with fresh amount of credits again.
I suppose we could either bump the initial credits or attempt to look for more peers if the current ones are drained.

**Comment by Tbaut** [comment]
Here is a quick overview of the [pip trace logs](https://gist.github.com/Tbaut/0e6d6839c8cea0d546e9598c8fee2da3).
- 11:41:12 the client is launched
- 11:41:43 the logs report a good average of ~4000+hdr/s for the last time
- Start to see a lot of peer disconnection 
- 11:41:48 the rate has dropped to ~700hdr/s
- 11:41:53 the rate has dropped to ~250hdr/s

Example of a good peer (peer 3) that gets limited within 12s:
```bash
cat pip.log| grep "peer 3\." | awk '{print $2"\t"$12" "$13"\t"$14" "$15 }'
11:41:14	Cost: 202189245;	Available: 30000000000
11:41:14	Cost: 202189245;	Available: 29797810755
11:41:14	Cost: 202189245;	Available: 29595621510
11:41:14	Cost: 202189245;	Available: 29393432265
11:41:14	Cost: 202189245;	Available: 29191243020
11:41:14	Cost: 202189245;	Available: 28989053775
11:41:14	Cost: 202189245;	Available: 28786864530
11:41:14	Cost: 202189245;	Available: 28584675285
11:41:14	Cost: 202189245;	Available: 28382486040
11:41:14	Cost: 202189245;	Available: 28180296795
11:41:14	Cost: 202189245;	Available: 27978107550
11:41:14	Cost: 202189245;	Available: 27775918305
11:41:15	Cost: 202189245;	Available: 27573729060
11:41:15	Cost: 202189245;	Available: 27371539815
11:41:15	Cost: 202189245;	Available: 27169350570
11:41:15	Cost: 202189245;	Available: 26967161325
11:41:15	Cost: 202189245;	Available: 26764972080
11:41:15	Cost: 202189245;	Available: 26562782835
11:41:15	Cost: 202189245;	Available: 26360593590
11:41:15	Cost: 202189245;	Available: 26158404345
11:41:15	Cost: 202189245;	Available: 25956215100
11:41:15	Cost: 202189245;	Available: 25754025855
11:41:15	Cost: 202189245;	Available: 25551836610
11:41:15	Cost: 202189245;	Available: 25349647365
11:41:15	Cost: 202189245;	Available: 25147458120
11:41:15	Cost: 202189245;	Available: 24945268875
11:41:15	Cost: 202189245;	Available: 24743079630
11:41:15	Cost: 202189245;	Available: 24540890385
11:41:15	Cost: 202189245;	Available: 24338701140
11:41:15	Cost: 202189245;	Available: 24136511895
11:41:15	Cost: 202189245;	Available: 23934322650
11:41:15	Cost: 202189245;	Available: 23732133405
11:41:15	Cost: 202189245;	Available: 23529944160
11:41:15	Cost: 202189245;	Available: 23327754915
11:41:15	Cost: 202189245;	Available: 23125565670
11:41:15	Cost: 202189245;	Available: 22923376425
11:41:15	Cost: 202189245;	Available: 22721187180
11:41:15	Cost: 202189245;	Available: 22518997935
11:41:15	Cost: 202189245;	Available: 22316808690
11:41:15	Cost: 202189245;	Available: 22114619445
11:41:15	Cost: 202189245;	Available: 21912430200
11:41:15	Cost: 202189245;	Available: 21710240955
11:41:15	Cost: 202189245;	Available: 21508051710
11:41:15	Cost: 202189245;	Available: 21305862465
11:41:15	Cost: 202189245;	Available: 21103673220
11:41:15	Cost: 202189245;	Available: 20901483975
11:41:15	Cost: 202189245;	Available: 20699294730
11:41:15	Cost: 202189245;	Available: 20497105485
11:41:15	Cost: 202189245;	Available: 20294916240
11:41:15	Cost: 202189245;	Available: 20092726995
11:41:15	Cost: 202189245;	Available: 19890537750
11:41:15	Cost: 202189245;	Available: 19688348505
11:41:15	Cost: 202189245;	Available: 19486159260
11:41:16	Cost: 202189245;	Available: 19283970015
11:41:16	Cost: 202189245;	Available: 19081780770
11:41:16	Cost: 202189245;	Available: 18879591525
11:41:16	Cost: 202189245;	Available: 18677402280
11:41:16	Cost: 202189245;	Available: 18475213035
11:41:16	Cost: 202189245;	Available: 18273023790
11:41:16	Cost: 202189245;	Available: 18070834545
11:41:16	Cost: 202189245;	Available: 17868645300
11:41:17	Cost: 202189245;	Available: 17666456055
11:41:17	Cost: 202189245;	Available: 17464266810
11:41:17	Cost: 202189245;	Available: 17262077565
11:41:17	Cost: 202189245;	Available: 17059888320
11:41:17	Cost: 202189245;	Available: 16857699075
11:41:17	Cost: 202189245;	Available: 16655509830
11:41:18	Cost: 202189245;	Available: 16453320585
11:41:18	Cost: 202189245;	Available: 16251131340
11:41:18	Cost: 202189245;	Available: 16048942095
11:41:18	Cost: 202189245;	Available: 15846752850
11:41:18	Cost: 202189245;	Available: 15644563605
11:41:18	Cost: 202189245;	Available: 15442374360
11:41:18	Cost: 202189245;	Available: 15240185115
11:41:18	Cost: 202189245;	Available: 15037995870
11:41:18	Cost: 202189245;	Available: 14835806625
11:41:18	Cost: 202189245;	Available: 14633617380
11:41:18	Cost: 202189245;	Available: 14431428135
11:41:19	Cost: 202189245;	Available: 14229238890
11:41:19	Cost: 202189245;	Available: 14027049645
11:41:19	Cost: 202189245;	Available: 13824860400
11:41:19	Cost: 202189245;	Available: 13622671155
11:41:19	Cost: 202189245;	Available: 13420481910
11:41:19	Cost: 202189245;	Available: 13218292665
11:41:19	Cost: 202189245;	Available: 13016103420
11:41:19	Cost: 202189245;	Available: 12813914175
11:41:19	Cost: 202189245;	Available: 12611724930
11:41:19	Cost: 202189245;	Available: 12409535685
11:41:19	Cost: 202189245;	Available: 12207346440
11:41:19	Cost: 202189245;	Available: 12005157195
11:41:19	Cost: 202189245;	Available: 11802967950
11:41:19	Cost: 202189245;	Available: 11600778705
11:41:20	Cost: 202189245;	Available: 11398589460
11:41:20	Cost: 202189245;	Available: 11196400215
11:41:20	Cost: 202189245;	Available: 10994210970
11:41:20	Cost: 202189245;	Available: 10792021725
11:41:20	Cost: 202189245;	Available: 10589832480
11:41:20	Cost: 202189245;	Available: 10387643235
11:41:20	Cost: 202189245;	Available: 10185453990
11:41:20	Cost: 202189245;	Available: 9983264745
11:41:20	Cost: 202189245;	Available: 9781075500
11:41:20	Cost: 202189245;	Available: 9578886255
11:41:20	Cost: 202189245;	Available: 9376697010
11:41:20	Cost: 202189245;	Available: 9174507765
11:41:20	Cost: 202189245;	Available: 8972318520
11:41:23	Cost: 202189245;	Available: 9470129275
11:41:23	Cost: 202189245;	Available: 9367940030
11:41:23	Cost: 202189245;	Available: 9265750785
11:41:23	Cost: 202189245;	Available: 9063561540
11:41:23	Cost: 202189245;	Available: 8861372295
11:41:24	Cost: 202189245;	Available: 8659183050
11:41:24	Cost: 202189245;	Available: 8456993805
11:41:24	Cost: 202189245;	Available: 8254804560
11:41:24	Cost: 202189245;	Available: 8052615315
11:41:24	Cost: 202189245;	Available: 7950426070
11:41:24	Cost: 202189245;	Available: 7748236825
11:41:24	Cost: 202189245;	Available: 7546047580
11:41:24	Cost: 202189245;	Available: 7343858335
11:41:24	Cost: 202189245;	Available: 7141669090
11:41:25	Cost: 202189245;	Available: 6939479845
11:41:25	Cost: 202189245;	Available: 6737290600
11:41:25	Cost: 202189245;	Available: 6535101355
11:41:25	Cost: 202189245;	Available: 6332912110
11:41:25	Cost: 202189245;	Available: 6130722865
11:41:25	Cost: 202189245;	Available: 5928533620
11:41:25	Cost: 202189245;	Available: 5726344375
11:41:25	Cost: 202189245;	Available: 5524155130
11:41:25	Cost: 202189245;	Available: 5321965885
11:41:25	Cost: 202189245;	Available: 5119776640
11:41:25	Cost: 202189245;	Available: 4917587395
11:41:25	Cost: 202189245;	Available: 4715398150
11:41:25	Cost: 202189245;	Available: 4513208905
11:41:25	Cost: 202189245;	Available: 4311019660
11:41:25	Cost: 202189245;	Available: 4108830415
11:41:25	Cost: 202189245;	Available: 3906641170
11:41:25	Cost: 202189245;	Available: 3704451925
11:41:25	Cost: 202189245;	Available: 3502262680
11:41:25	Cost: 202189245;	Available: 3300073435
11:41:25	Cost: 202189245;	Available: 3097884190
11:41:25	Cost: 202189245;	Available: 2895694945
11:41:25	Cost: 202189245;	Available: 2693505700
11:41:25	Cost: 202189245;	Available: 2491316455
11:41:25	Cost: 202189245;	Available: 2289127210
11:41:25	Cost: 202189245;	Available: 2086937965
11:41:25	Cost: 202189245;	Available: 1884748720
11:41:26	Cost: 202189245;	Available: 1682559475
11:41:26	Cost: 202189245;	Available: 1480370230
11:41:26	Cost: 202189245;	Available: 1278180985
11:41:26	Cost: 202189245;	Available: 1075991740
11:41:26	Cost: 202189245;	Available: 873802495
11:41:26	Cost: 202189245;	Available: 671613250
11:41:26	Cost: 202189245;	Available: 469424005
11:41:26	Cost: 202189245;	Available: 267234760
11:41:33	Cost: 202982144;	Available: 865045515
11:41:33	Cost: 202189245;	Available: 662063371
11:41:33	Cost: 202189245;	Available: 559874126
11:41:33	Cost: 202189245;	Available: 357684881
11:41:34	Cost: 202189245;	Available: 255495636
11:41:37	Cost: 202189245;	Available: 253306391
11:41:39	Cost: 202189245;	Available: 251117146
11:41:42	Cost: 202189245;	Available: 248927901
11:41:42	Cost: 202189245;	Available: 246738656
11:41:48	Cost: 202189245;	Available: 544549411
11:41:48	Cost: 202189245;	Available: 342360166
11:41:48	Cost: 202189245;	Available: 240170921
11:41:53	Cost: 202189245;	Available: 437981676
11:41:53	Cost: 202189245;	Available: 235792431
11:41:56	Cost: 202189245;	Available: 333603186
11:41:57	Cost: 202189245;	Available: 231413941
11:42:03	Cost: 202189245;	Available: 629224696
11:42:03	Cost: 202189245;	Available: 427035451
11:42:03	Cost: 202189245;	Available: 324846206
11:42:08	Cost: 202189245;	Available: 522656961
11:42:08	Cost: 202189245;	Available: 320467716
11:42:08	Cost: 202189245;	Available: 218278471
```
